File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void solve(int i,int j,vector<string>&v,string s) 
    {
        if (i==0 && j==0)
        {
            v.push_back(s);
            return;
        }
        
        if (i>0) solve(i-1,j,v,s+'(');
        if (j>0 && j>i) solve(i,j-1,v,s+')');
    }
    
    vector<string> generateParenthesis(int n) {
        vector<string>v;
        solve(n, n,v,"");
        return v; 
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        l=[]
        def fun(i,j,n,s):
            if i==j and i+j==n*2:
                l.append(s)
                return 
            if i<n:
                fun(i+1,j,n,s+'(')
            if j<i:
                fun(i,j+1,n,s+')')
        fun(0,0,n,"")
        return l 
                